src/implicit_function.jl

```diff
@@ -30,39 +30,47 @@ function ImplicitFunction(forward::F, conditions::C) where {F,C}
     return ImplicitFunction(forward, conditions, gmres)
 end
 
-struct SolverFailureException <: Exception
+struct SolverFailureException{S} <: Exception
     msg::String
+    stats::S
+end
+
+function Base.show(io::IO, sfe::SolverFailureException)
+    println(io, "SolverFailureException: $(sfe.msg) \n Solver stats: $(sfe.stats)")
 end
 
 """
-    implicit(x)
+    implicit(x[; kwargs...])
 
 Make [`ImplicitFunction{F,C,L}`](@ref) callable by applying `implicit.forward`.
 """
-(implicit::ImplicitFunction)(x) = implicit.forward(x)
+(implicit::ImplicitFunction)(x; kwargs...) = implicit.forward(x; kwargs...)
 
 """
-    frule(rc, (_, dx), implicit, x)
+    frule(rc, (_, dx), implicit, x[; kwargs...])
 
 Custom forward rule for [`ImplicitFunction{F,C,L}`](@ref).
 
 We compute the Jacobian-vector product `Jv` by solving `Au = Bv` and setting `Jv = u`.
+Keyword arguments are given to both `implicit.forward` and `implicit.conditions`.
 """
 function ChainRulesCore.frule(
-    rc::RuleConfig, (_, dx), implicit::ImplicitFunction, x::AbstractArray{R}
+    rc::RuleConfig, (_, dx), implicit::ImplicitFunction, x::AbstractArray{R}; kwargs...
 ) where {R<:Real}
-    (; forward, conditions, linear_solver) = implicit
+    forward = implicit.forward
+    conditions = implicit.conditions
+    linear_solver = implicit.linear_solver
 
-    y = forward(x)
+    y = forward(x; kwargs...)
 
-    conditions_x(x̃) = conditions(x̃, y)
-    conditions_y(ỹ) = -conditions(x, ỹ)
+    conditions_x(x̃; kwargs...) = conditions(x̃, y; kwargs...)
```

Comment:
I don't think conditions should have kwargs in general. What's a use case for this?

Reply:
Ah I see it used in the OT example, that's clever. You are using kwargs to indicate parameters we don't want to differentiate wrt? Is this consistent with how ChainRules handles kwargs though?

Reply:
I think it's fine even if not, but we should document it.

Reply:
ChainRules does not differentiate wrt kwargs, so I think we're good

Reply:
But I'd like your input on a type inference problem I found with ChainRulesTestUtils:

- on lines 107-108, `kwargs` is free, it refers to an argument of the closure
- on lines 110-111, `kwargs` is not free, it refers to the keyword arguments passed to the `rrule`

This is a workaround I found to avoid type inference errors. The natural thing to do was to pass the keyword arguments from the `rrule` directly in lines 107-108, but for some reason this gave rise to an inference error in the first round of tests (unconstrained optimization). Any clue why? Is that a problem at all?

Reply:
could you commit the form that gives an inference error to another branch so I know we are talking about the same thing?

Reply:
https://github.com/gdalle/ImplicitDifferentiation.jl/commit/2232ef10b10ffebba0f48efe12de17a0c1a9d8c0

Reply:
I really think we're hitting some quirk about closures that has no obvious solution

Reply:
Seems to be a problem with the RuleConfig used by ChainRulesTestUtils. I reduced the inference issue to:
```julia
using Test, ChainRulesTestUtils

struct FF{F, Y, K}
	f::F
	y::Y
	kwargs::K
end
(f::FF)(x) = f.f(x, f.y; f.kwargs...)

zero_gradient(x, y) = 2(y - x);
f = FF(zero_gradient, rand(3), NamedTuple())
rc = ChainRulesTestUtils.TestConfig()
x = rand(3)

@inferred rrule_via_ad(rc, f, x)
```

Reply:
Reduced further to:
```julia
using Test, ChainRulesCore, ChainRulesTestUtils

struct FF{F, Y}
	f::F
	y::Y
end
(f::FF)(x) = f.f(x, f.y)

g(x, y) = x + y;
f = FF(g, rand(3))
rc = ChainRulesTestUtils.TestConfig()
x = rand(3)

@inferred rrule_via_ad(rc, f, x)
```

Reply:
perhaps worth opening an issue there

Reply:
https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/246

Reply:
merging this for now until the bug is fixed upstream